ALE with finite volume for fluid flow

AlbertJP7 opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
The AMMT program expects us to do ALE with finite volumes for the fluid flow and CFEM for the displacements

## Design
- Significant development to enable finite volume on displaced mesh/problem
- Add `NodeArg` for functor evaluation
- Implement `ElemArg` for FE for coupling the mesh velocity into the convective term of the momentum and energy transport equations
- Implement Qp evaluations for FV by just passing to `ElemPointArg` overload which does a two term expansion

## Impact
ALE simulation capability with finite volume modeling the NS equations and nodal displacements for the displacing mesh

AlbertJP7 committed:
Node Argument header functor files

Refs #19420

AlbertJP7 committed:
Changes to AMMT from NodeFunctorArgs

AlbertJP7 committed:
Changes made to NodeFunctorArgs

AlbertJP7 committed:
Edited files for NodeArg to maybe work

AlbertJP7 committed:
Other files

lindsayad commented:
@AlbertJP7 I forgot to tell you to add an issue reference in at least one of your commits. You could run
```
git commit --amend
```
and then in your text editor add to the commit message "refs https://github.com/idaholab/moose/issues/19420"

GiudGiud commented on framework/include/base/MooseFunctor.h:
```diff
@@ -252,6 +257,10 @@ class FunctorBase : public FunctorAbstract
 
+  virtual GradientType evaluateGradient(const NodeArg &, const StateArg &) const
+  {
+    mooseError("Element gradient not implemented for functor " + functorName());
```
```suggestion
    mooseError("Gradient at node not implemented for functor " + functorName());
```

GiudGiud commented on framework/include/base/MooseFunctor.h:
```diff
@@ -300,6 +309,11 @@ class FunctorBase : public FunctorAbstract
+
+  virtual DotType evaluateDot(const NodeArg &, const StateArg &) const
+  {
+    mooseError("Element time derivative not implemented for functor " + functorName());
```
```suggestion
    mooseError("Time derivative at node not implemented for functor " + functorName());
```

GiudGiud commented on framework/include/base/MooseFunctor.h:
```diff
@@ -369,6 +383,10 @@ class FunctorBase : public FunctorAbstract
   /// Map from face arguments to their cached evaluations
   mutable std::map<FaceArg, ValueType> _face_arg_to_value;
 
+  /// Map from element arguments to their cached evaluations
```
```suggestion
  /// Map from nodal arguments to their cached evaluations
```

GiudGiud commented on framework/include/variables/MooseVariableFV.h:
```diff
@@ -714,6 +717,27 @@ MooseVariableFV<OutputType>::evaluate(const ElemArg & elem_arg, const StateArg &
+    sum += getElemValue(elem_arg, state);
+    ++num_values;
+  }
+  return sum / num_values;
```
not volume weighted?
what is our current use for this?

lindsayad replied:
I think something like an inverse distance weighting would be more appropriate. The use for this and the biggest driver for this PR is being able to construct a velocity field at nodes for coupling with surface Dirichlet BCs for displacements in ALE simulations where the surface is displacing

GiudGiud replied:
Sounds good. Whatever we use we should strive to have something that is centered/second order in the default case (seems good enough) and document it

lindsayad replied:
For second order accuracy on boundary nodes, the evaluation would probably need to consider the adjacent cell center gradients to extrapolate

GiudGiud replied:
I think so. those would be considered indirectly if we do an average of face centered values (instead of element-centered values)

lindsayad replied:
That is true. Quite a few more operations unfortunately. Maybe weighting of face-based evaluations should be the default for boundary nodes, and element-based should be the default for internal nodes. (Of course then there is the question of inexpensively determining if you're a boundary node)

GiudGiud replied:
I like that!
For weighting schemes, I am happy with volume-weighting or inverse-distance weighting.
@grmnptr maybe you know what is more common in FV? 
In FE I dont know either!

GiudGiud replied:
nevermind this isnt for FV :D

GiudGiud commented on framework/include/functions/Function.h:
```diff
@@ -164,6 +165,7 @@ class Function : public MooseObject,
                      const Moose::StateArg & state) const override final;
   ValueType evaluate(const ElemPointArg & elem_point,
                      const Moose::StateArg & state) const override final;
+  ValueType evaluate(const NodeArg & node, const Moose::StateArg & state) const override final;
```
is the implementation missing?

GiudGiud reviewed (changes requested):
this is going to be such a strong capability

lindsayad committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1767757/) on 14b1df8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25181/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25181/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1acda23805be34b178665b209b2988bba734b183`

lindsayad committed:
clang-format

GiudGiud commented on framework/include/base/MooseFunctorArguments.h:
```diff
@@ -132,6 +132,12 @@ struct FaceArg
   }
 };
 
+struct NodeArg
```
docstrings

GiudGiud commented on framework/include/base/MooseFunctorArguments.h:
```diff
@@ -132,6 +132,12 @@ struct FaceArg
+struct NodeArg
+{
+  const Node * node;
+  SubdomainID subdomain_id;
```
the bare minimum, nice

GiudGiud commented on modules/navier_stokes/include/utils/CellCenteredMapFunctor.h:
```diff
@@ -209,6 +211,33 @@ CellCenteredMapFunctor<T, Map>::evaluate(const ElemSideQpArg &, const StateArg &
+
+  try
+  {
+    return libmesh_map_find(*this, node->id());
```
so the cellCenteredMapFunctor has node-cached values?
seems odd? I feel like node value of a cellCentered functor would be some sort of average of cell-centered values?

if you want a map then it would be a NodeMapFunctor ?

GiudGiud reviewed (commented):
got an email, took a look. I know this is a draft, I mean no trouble

lindsayad committed:
Fix compilation errors/warnings

lindsayad committed:
Change to use functors in INSAD*DisplaceBoundary objects

lindsayad committed:
Implement NodeArg for MooseVariableFE

lindsayad committed:
Help dumb compilers

lindsayad committed:
Add associated_subdomain to laser input

lindsayad committed:
adding Function functors can be simpler now

lindsayad committed:
Create INSADEnergyMeshConvection

lindsayad committed:
INSADMeshConvection -> INSADMomentumMeshAdvection

lindsayad committed:
INSADEnergyMeshConvection -> INSADEnergyMeshAdvection

lindsayad committed:
Add dot functor material properties to GenericVectorFunctorMaterial

lindsayad committed:
Add mesh convection treatment to FV kernels

To give consistent treatment around material time derivatives I
also had to create new `WCNSFVMassAdvection` and `PWCNSFVMassAdvection`
classes.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1771982/) on 1c8a577 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25181/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25181/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4fc72f167067e0185d4154811a790f2108476472`

lindsayad committed:
Fix bug in 3d-rc-no-slip input

lindsayad committed:
Add use_displaced_mesh = true for remaining FE laser welding mats

This doesn't affect any of the current simulation results but it
could have if we were using `_grad_k` somewhere since although
the temperature values are not functions of the mesh displacement,
the temperature gradients are

lindsayad committed:
Create FVFunctorRadiativeBC

lindsayad committed:
Create FVGaussianEnergyFluxBC

lindsayad committed:
Building our FV ALE input

lindsayad committed:
Fixes to allow FV on displaced meshes

lindsayad committed:
Create INSFVVaporRecoilPressureMomentumFlux

lindsayad committed:
Create functor version of Aria material

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1772630/) on 6399ede wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25181/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25181/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4fc72f167067e0185d4154811a790f2108476472`

lindsayad committed:
Implement some missing variable functor evaluation routines

- ElemArg for finite elements
- ElemQpArg and ElemSideQpArg for finite volumes

lindsayad committed:
Implement evaluateDot with ElemArg for FE

lindsayad committed:
Do separate loops for displaced FV objects

lindsayad committed:
Duplicate functor materials for undisp and disp problems

lindsayad committed:
Make sure to create node functor in PWBBLF

lindsayad committed:
Explicit signed char for AttribDisplaced key

lindsayad committed:
Actually signal an exception in ComputeFVFluxThread

lindsayad committed:
When running displaced, make sure to prepare displaced data

lindsayad committed:
Use structured binding for clarity in buildFiniteVolumeInfo

lindsayad committed:
Make sure displaced elem gets reinit'd for FV elemental kernels

lindsayad committed:
Require qp computations for FVRadiativeHeatFluxBCBase

lindsayad committed:
Constrain away pressure nullspace

lindsayad committed:
Change to creating disp dot functor in the RC UO

lindsayad committed:
Implement FaceArg for MooseVariableFE::evaluateDot

lindsayad committed:
Try evaluating NodeArg for FV with two term expansion

I thought this would better match the sparsity pattern coming out
of the dummy integrated BC (which would also use a two term expansion
to the boundary face) but I'm still getting new nonzero errors

lindsayad committed:
Create 2D FV input

lindsayad committed:
Do all caching of residuals/Jacobians through subproblem

This appears to fix the sparsity pattern issues I was running into
for a reason I have already forgotten

lindsayad committed:
Update elem for mesh adaptivity in FVPointValueConstraint

lindsayad committed:
No need to cacheJacobian; no FV objects use _local_ke

lindsayad committed:
Mark finite volume info as dirty after displacing mesh

lindsayad committed:
Dont scale vels

lindsayad committed:
Move constraint point into middle of domain

lindsayad committed:
Try out INSFVMomentumPressureFlux

lindsayad committed:
Create INSFVMomentumPressureFlux

lindsayad committed:
Disable adaptivity

lindsayad committed:
use fine initial mesh

lindsayad committed:
Decrease solid phase viscosity for less discontinuity

lindsayad committed:
Revert "Try out INSFVMomentumPressureFlux"

This reverts commit 2dcc47a802347bdaf0804ff81e945e57c42c6a8e.

lindsayad committed:
Make sure to setup displaced variables

lindsayad committed:
disp_x shouldn't be required in the interpolator

lindsayad committed:
Futzing around with the 2D FV input

Something still doesn't seem quite right. Convergence is terrible
and I'm not sure the physics look right

lindsayad committed:
Add a great looking 2D finite element input

All the profiles look like I would hope. Most notably the vel-x
points away from the impinging spot whereas for the FV sim the
vel-x is pointing towards the spot. That definitely doesn't seem
right

lindsayad committed:
Add channel flow displaced mesh test

lindsayad committed:
Add displaced lid-driven test

lindsayad committed:
Use one term expansion for FV at nodes

lindsayad committed:
Revert "Use one term expansion for FV at nodes"

This reverts commit d22adab8dc5971efef08e6c0efe0d07dc7536151.

lindsayad committed:
A good result. Hoozaw

lindsayad committed:
Good result on a bit finer mesh

lindsayad committed:
An awesome sim

lindsayad committed:
Remove unimportant lines in FV input

lindsayad committed:
Add 2d-fv test

lindsayad commented:
@GiudGiud are you open to me opening a new PR under my name? I've invited @AlbertJP7 to our moose-users group but he hasn't accepted so I have to manually trigger testing every time I push. It's only mildly obnoxious so if you want to keep the PR here, then we can do that

GiudGiud commented:
no problem

I actually did that for all my interns' work to avoid this

lindsayad commented:
You have more foresight than I 😄

lindsayad commented:
Continuing in https://github.com/idaholab/moose/pull/25574
